### ISSUE TITLE
SCT's multithreading uses all of our poor clusters' CPU resources

### ISSUE THREAD

**Issue by taowa** [spinalcordtoolbox/spinalcordtoolbox#3620]
As you can see in the attached image, sometimes SCT decides that all of a machine's CPUs belong to it.

![Screenshot 2021-12-08 at 16 51 12](https://user-images.githubusercontent.com/1357889/145290856-57fbdee0-627c-4034-b29c-5fef3c57e99c.png)

Ideally, we'd use a percentage of the CPUs, which would be below 100% unless otherwise specified. Otherwise, (number of CPUs - 2) is probably(?) fine as well.

I think this is set here:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/8985ba391ff149cb7af8a086dd7d0d28be90f010/spinalcordtoolbox/scripts/sct_run_batch.py#L428

but I have no clue.

This is an issue for us, as it makes logging into clusters reallllly slow, among other issues. I figure our friends at Compute Canada may also want us to be a bit careful.

 \- tr

**Comment by joshuacwnewton** [comment]
Just to make sure I understand what I'm seeing in the screenshot... It mentions:

```
registration_evaluation_models/bids_registration.py --model-path [...]
```

From a GitHub search, it looks like this is referring to @EvanBeal's work from https://github.com/ivadomed/multimodal-registration/pull/3. 
(Specifically, [L77 from `pipeline_bids_register_evaluate.sh`](https://github.com/ivadomed/multimodal-registration/blob/baef60fc9122d98992add5c02873303da0045053/pipeline_bids_register_evaluate.sh#L77)):

```
python $PATH_SCRIPT/bids_registration.py --model-path $PATH_SCRIPT/model/$REGISTRATION_MODEL --fx-img-path $file_t1_before_proc --mov-img-path $file_t2_before_proc --fx-img-contrast T1w
```

**Comment by joshuacwnewton** [comment]
> I think this is set here:
> 
> https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/8985ba391ff149cb7af8a086dd7d0d28be90f010/spinalcordtoolbox/scripts/sct_run_batch.py#L428

The line you highlighted is indeed relevant, but I think it needs a little more context to highlight the conditional that surrounds it:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/8985ba391ff149cb7af8a086dd7d0d28be90f010/spinalcordtoolbox/scripts/sct_run_batch.py#L426-L429

`arguments.jobs` corresponds to the `-jobs` argument to `sct_run_batch`. In theory, it _should_ be 1 by default:

```
  -jobs <int>           The number of jobs to run in parallel. Either an integer
                        greater than or equal to one specifying the number of
                        cores, 0 or a negative integer specifying number of
                        cores minus that number. For example '-jobs -1' will run
                        with all the available cores minus one job in parallel.
                        Set '-jobs 0' to use all available cores.
                        This argument enables process-based parallelism, while
                        '-itk-threads' enables thread-based parallelism. You may
                        need to tweak both to find a balance that works best for
                        your system. (default: 1)

```

So, just to double-check -- Have you used `-jobs -1` or `-jobs 0` by chance, @EvanBeal ?

**Comment by EvanBeal** [comment]
Yes indeed, it's from my current experiments on our CPUs. And yes in one of my call I used `sct_run_batch` with the argument `-jobs -1`. So, I think the screenshot was taken from this situation.

**Comment by joshuacwnewton** [comment]
Huzzah! Thanks so much for the quick response, @EvanBeal! :heart: 

The good news, then, is that SCT doesn't use all of the CPU's resources by default. :smile: 

----

> Ideally, we'd use a percentage of the CPUs, which would be below 100% unless otherwise specified.

@taowa Are there perhaps improvements we could make to SCT's documentation and/or our "computing resources" page? I'm not exactly sure what sort of recommendations to make here... is there a meaningful difference between 75%/80%/95%/etc. in terms of the issues you describe?

**Comment by EvanBeal** [comment]
To investigate this issue further, I ran the following script `test_cpu.sh` on a dataset (BIDS convention) of 6 subjects with `sct_run_batch` and specifying `6` as the number of jobs to run in parallel. The command line is: `sct_run_batch -jobs 6 -path-data bids_dataset -path-out res -script test_cpu.sh`.

<details>
  <summary>test_cpu.sh</summary>
  
```
#!/bin/bash
#
# Usage:
# sct_run_batch -jobs 1 -path-data bids_dataset -path-out res -script test_cpu.sh
#
# Author: Evan Béal

set -x
# Immediately exit if error
set -e -o pipefail

# Exit if user presses CTRL+C (Linux, OSX)
trap "echo Caught Keyboard Interrupt within script. Exiting now.; exit" INT

# Retrieve input params
SUBJECT=$1

# Save script path
PATH_SCRIPT=$PWD

# get starting time:
start=`date +%s`

# FUNCTIONS
# ==============================================================================

# Perform segmentation.
segment(){
  local file="$1"
  local contrast="$2"
  folder_contrast="anat"

  echo "Proceeding with automatic segmentation"
  # Segment spinal cord
  sct_deepseg_sc -i ${file}.nii.gz -c $contrast -qc ${PATH_QC} -qc-subject ${SUBJECT}
}

# SCRIPT STARTS HERE
# ==============================================================================
# Display useful info for the log, such as SCT version, RAM and CPU cores available
sct_check_dependencies -short

SUBJECT_ID=$(dirname "$SUBJECT")
SES=$(basename "$SUBJECT")
# Go to folder where data will be copied and processed
cd ${PATH_DATA_PROCESSED}
# Copy source images
mkdir -p ${SUBJECT}
rsync -avzh $PATH_DATA/$SUBJECT/ ${SUBJECT}
# Go to anat folder where all structural data are located
echo $PWD
cd ${SUBJECT}/anat/

file_t1_before_proc="${SES}_T1w"
file_t2_before_proc="${SES}_T2w"

# Segment spinal cord
segment $file_t1_before_proc "t1"
segment $file_t2_before_proc "t2"

file_t1_seg="${SES}_T1w_seg"
file_t2_seg="${SES}_T2w_seg"

# Generate QC report to assess registration
sct_qc -i ${file_t1_before_proc}.nii.gz -s ${file_t1_seg}.nii.gz -d ${file_t2_before_proc}.nii.gz -p sct_register_multimodal -qc ${PATH_QC} -qc-subject ${SUBJECT}

# Verify presence of output files and write log file if error
# ------------------------------------------------------------------------------
FILES_TO_CHECK=(
  "${PATH_DATA_PROCESSED}/${SUBJECT}/anat/${SES}_T1w.nii.gz"
  "${PATH_DATA_PROCESSED}/${SUBJECT}/anat/${SES}_T2w.nii.gz"
  "${PATH_DATA_PROCESSED}/${SUBJECT}/anat/${SES}_T1w_seg.nii.gz"
  "${PATH_DATA_PROCESSED}/${SUBJECT}/anat/${SES}_T2w_seg.nii.gz"
)
pwd
for file in ${FILES_TO_CHECK[@]}; do
  if [[ ! -e $file ]]; then
    echo "${file} does not exist" >> $PATH_LOG/_error_check_output_files.log
  fi
done

# Display useful info for the log
end=`date +%s`
runtime=$((end-start))
echo
echo "~~~"
echo "SCT version: `sct_version`"
echo "Ran on:      `uname -nsr`"
echo "Duration:    $(($runtime / 3600))hrs $((($runtime / 60) % 60))min $(($runtime % 60))sec"
echo "~~~"
```
</details>

What I observed with this experiment is that at the beginning of the process, there is indeed only 6 CPUs that are used:

<img width="1506" alt="Capture d’écran 2021-12-09 à 12 06 59" src="https://user-images.githubusercontent.com/32447627/145443939-1ef32e14-70f7-4bd7-9eb6-5700a629952b.png">

However, after some time the python script (here `sct_deepseg_sc.py`) will have no limitation anymore on the number of CPUs to use, resulting on all the CPUs of the machine being used and this latter being slow for other tasks.

<img width="1522" alt="Capture d’écran 2021-12-09 à 12 08 52" src="https://user-images.githubusercontent.com/32447627/145444349-132f475c-dcad-4364-b7f3-18d4f7222e92.png">

Therefore, it seems that the number of CPUs specified in `sct_run_batch` controls well the number of CPUs used for the shell and bash commands in the script but not necessarily for the python scripts as it seems that they have access to all the CPUs for some computations, even though there should be a restriction on the number of CPUs to use.

**Comment by jcohenadad** [comment]
> However, after some time the python script (here sct_deepseg_sc.py) will have no limitation anymore on the number of CPUs to use, resulting on all the CPUs of the machine being used and this latter being slow for other tasks.

Indeed, I've also noticed this several times and I've always wanted to start a brainstorming session on this. I'm glad the discussion thread is now open 😊

**Comment by joshuacwnewton** [comment]
> **Disclaimer:** I'm still a bit new to multiprocessing, so a lot of what I've written here was stitched together from cursory searches and skimming various resources. If there's anything I get wrong here, please call me out on it! 

I've done a little digging, and here's my best attempt to summarize the problem:

  * <details> 
    <summary> Overview of how <code>sct_run_batch</code> operates internally (click to open) </summary>
    
    1. **Python layer 1:** [`multiprocessing.Pool(jobs)`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/685ab17ff9bebeb38228d52a55b2507b60b52446/spinalcordtoolbox/scripts/sct_run_batch.py#L441) -> [`pool.imap()`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/685ab17ff9bebeb38228d52a55b2507b60b52446/spinalcordtoolbox/scripts/sct_run_batch.py#L453) -> N worker processes, each distributed to 1 subject directory at a time
        
        > **Note:** I'm assuming the intent of using use of `multiprocessing` here is to overcome the fact the "in Python, single-CPU use is caused by the global interpreter lock (GIL), which allows only one thread to carry the Python interpreter at any given time. The GIL was implemented to handle a memory management issue, but as a result, Python is limited to using a single processor." [(source)](https://urban-institute.medium.com/using-multiprocessing-to-make-python-code-faster-23ea5ef996ba)
    
    2. Then, for each worker process:
        i. **Worker process layer:**  [`run_single()`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/685ab17ff9bebeb38228d52a55b2507b60b52446/spinalcordtoolboxscripts/sct_run_batch.py#L157) -> [`subprocess.run(script.sh)`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/685ab17ff9bebeb38228d52a55b2507b60b52446/spinalcordtoolbox/scripts/sct_run_batch.py#L208-L211)
        ii. **Subprocess/shell script layer:** Runs commands, including `python` scripts and SCT CLI scripts (e.g. `sct_deepseg_sc`)
        iii. **Python layer 2:** `sct_deepseg_sc` -> scientific Python libraries (`numpy`, `tensorflow`, etc.) -> C code
        iv. **Low-level C code**:  Unrestricted by the limitations of Python's GIL:
     
        > Python alone will only use one core, although scientific libraries such as numpy can execute the logic at a lower level (using c bindings) and can use all core and/or your GPU. [(source)](https://www.reddit.com/r/Python/comments/4kqb4c/does_python_use_all_cores_of_the_processor/d3gxqt0/)
    
        > Tensorflow automatically runs the computations on as many cores as are available on a single machine. [(source)](https://stackoverflow.com/a/41597238)
    
    </details>

tl;dr: 

* The specific control we have in place (`-jobs` + `multiprocessing`) is one approach to get around the GIL and gain multi-core concurrency for Python programs. 
* Yet, the Python libraries we use (`numpy`, `tensorflow`, etc.) have their own ways of getting around the GIL and gaining multi-core concurrency (by utilising low-level C code). This effectively sidesteps the CPU limit that SCT introduces through `-jobs`.

Here are some closing thoughts I have based on the above:

* How useful even is `-jobs`? If our scripts were pure Python, then using `-jobs` with `multiprocessing` would make a lot of sense. But, given that the most resource-intensive parts of our scripts _aren't_ in Python, how much of a benefit does it really offer? Have we done any benchmark comparisons between `-jobs 1` and `-jobs N`?
* I don't immediately have any ideas on imposing CPU limits on the low-level parts of our scripts.
* As an alternative, one suggestion I found during my research was [to use `nice`](https://stackoverflow.com/a/42130713) to lower the priority of the subprocesses. While that wouldn't limit the CPU usage, it might mitigate the problems observed by @taowa. 
  
    > You can `nice` the subprocesses. This way, though they will still eat 100% of the CPU, when you start other applications, the OS gives preference to the other applications.
  
* However, given how many layers of separation there are between `sct_run_batch` and the low-level C code that eventually gets run, I have no idea if the benefits of `nice` will trickle down through the layers to where we need it to be.

**Comment by jcohenadad** [comment]
> Have we done any benchmark comparisons between -jobs 1 and -jobs N?

Yes, I have (and I know some students also did in the past), and there is _definitely_ a difference between 1 and N. I know there is a GH discussion thread somewhere with htop and timing comparisons (maybe on SCT repos?). I think @hermancollin and @zougloub were doing some benchmarking on this. However the sweet spot depends on the processes run, the number of CPU cores available, the size of the data we are dealing with, the available RAM, etc..

**Comment by EvanBeal** [comment]
From what I have seen and as mentioned by @joshuacwnewton (_Tensorflow automatically runs the computations on as many cores as are available on a single machine_), the problem seems to come (mainly) from TensorFlow (TF) (at least in my case), so a potential solution is to force TF to run on only one (or a limited number of) CPU thread. This seems to be possible by adding the following code at the beginning of the `.py` scripts that use a TF model. 

For TF 1.x
```
session_conf = tf.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
sess = tf.Session(config=session_conf)
```

Or for TF 2.x
```
session_conf = tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
sess = tf.compat.v1.Session(config=session_conf)
```

Therefore, for each worker process, when TF will be used in the Python layer, it will have access to only 1 (or the specified number) CPU.

This solution is far from ideal because it limits the number of CPUs used by TF in an inelegant/unflexible way, but it may be useful to start dealing with this problem.

Considering the same example as my previous comment (with `test_cpu.sh`), adding the code specified above to the file `sct_deepseg_sc.py` does not totally lead to a process using only the number of CPUs specified but my guess (and what is suggested by the following images) would be that this is due to the part linked to the binaries (with `isct_spine_detect`).

<details>
  <summary>When the binary `isct_spine_detect` appears</summary>

<img width="1432" alt="Capture d’écran 2021-12-09 à 18 19 06" src="https://user-images.githubusercontent.com/32447627/145491839-68b435fb-d116-450d-a71a-ef77d3937bb7.png">

</details>

<details>
  <summary>Just a few seconds later</summary>

<img width="1436" alt="Capture d’écran 2021-12-09 à 18 19 14" src="https://user-images.githubusercontent.com/32447627/145491942-34e9d3da-34b9-4971-973c-70be1c93aab4.png">

</details>

Therefore, to ensure that this could be a solution to our issue, I tested to add these lines in the file where the problem was first observed `bids_registration.py` that is part of the shell script `pipeline_bids_register_evaluate.sh` and as we can see on the two screenshots, it worked perfectly to reduce the number of CPUs used by this process (only one CPU per subject was used). This test was done on 6 subjects and using `sct_run_batch -jobs 6 -path-data bids_dataset -path-out res -script pipeline_bids_register_evaluate.sh`.

<details>
  <summary>Without specifying the number of CPUs for TF</summary>

<img width="1443" alt="Capture d’écran 2021-12-09 à 18 18 07" src="https://user-images.githubusercontent.com/32447627/145492081-80420c7d-87ef-4647-827d-f2774b071344.png">

</details>

<details>
  <summary>When specifying the number of CPUs for TF</summary>

<img width="1449" alt="Capture d’écran 2021-12-09 à 18 04 08" src="https://user-images.githubusercontent.com/32447627/145492164-51a231fa-dc53-4a9f-b102-91ed3379704d.png">

</details>

Additional comment: running `pipeline_bids_register_evaluate.sh` on 6 subjects took 3 min 36 sec when all the CPUs (128) were used by TF, whereas when specifying to use 1 CPU per subject, it took 7 min 18 sec. In both cases, the number of jobs in `sct_run_batch` was set to 6. This indicates that indeed less computational resource was used when adding the lines mentioned above.

**Comment by joshuacwnewton** [comment]
> <details>
>   <summary>Without specifying the number of CPUs for TF</summary>
> 
> <img width="1443" alt="Capture d’écran 2021-12-09 à 18 18 07" src="https://user-images.githubusercontent.com/32447627/145492081-80420c7d-87ef-4647-827d-f2774b071344.png">
> 
> </details>
> 
> <details>
>   <summary>When specifying the number of CPUs for TF</summary>
> 
> <img width="1449" alt="Capture d’écran 2021-12-09 à 18 04 08" src="https://user-images.githubusercontent.com/32447627/145492164-51a231fa-dc53-4a9f-b102-91ed3379704d.png">
> 
> </details>

These experiments are very interesting! Even as a rough proof of concept, it's exciting to see `-jobs` working as it was (theoretically) intended to work.

> Additional comment: running `pipeline_bids_register_evaluate.sh` on 6 subjects took 3 min 36 sec when all the CPUs (128) were used by TF, whereas when specifying to use 1 CPU per subject, it took 7 min 18 sec. In both cases, the number of jobs in `sct_run_batch` was set to 6. This indicates that indeed less computational resource was used when adding the lines mentioned above.

For what it's worth, this was still just a toy example using 6 subjects and 6 jobs. On a larger dataset, where the number of subjects is closer to the number of available CPU cores, `jobs -1` (or even `jobs -10`, etc.) would become feasible again, and we could use a larger % of the available CPUs. In that case, I imagine the two different times would be a bit closer? :thinking:

----

I have one more (pretty rough) idea, but... What if we set a new environment variable using `sct_run_batch` called `NUMBER_OF_JOBS`, similar to what we do elsewhere:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/685ab17ff9bebeb38228d52a55b2507b60b52446/spinalcordtoolbox/scripts/sct_run_batch.py#L195-L204

Then, depending on the number of jobs (i.e. number of concurrent subjects) requested, we change TF's behavior?

* If `jobs 1` was passed (i.e. the default behavior), then we know the user intends to process a single subject at a time, and therefore we keep the default TF behavior of managing its own thread usage.
* If `jobs >1` was passed, then we can assume that the user is working on a larger dataset, and therefore they can spread all of the subjects out across the different CPU cores, and 1 CPU per TF instance would be desirable.

**Comment by jcohenadad** [comment]
> Then, depending on the number of jobs requested (i.e. number of concurrent subjects to be processed), we change TF's behavior?

👍

**Comment by EvanBeal** [comment]
> For what it's worth, this was still just a toy example using 6 subjects and 6 jobs, so the time difference might not even be that bad? On a larger dataset (or a less capable CPU), where the number of subjects is closer to the number of available CPU cores, `jobs -1` would become feasible again, and we could use a larger % of the available CPUs. In that case, I imagine the two different times would be a bit closer? 🤔

Yes I think you're right ! The larger the number of processes to run in parallel the closer the times in the two scenarios should be normally.

---

> although we need to keep in mind that TF will 'soon' be deprecated, so we should also consider the time it would take to do this implementation vs. the benefits it would bring until TF gets deprecated.

Yes I agree, it may not be worth spending too much time on it. However, I think that a similar behavior is (will probably be) observed with PyTorch. And it seems that it can be solved in a similar manner by specifying the number of threads to use before calling other PyTorch functions (see [`torch.set_num_threads(int)`](https://pytorch.org/docs/stable/generated/torch.set_num_threads.html#torch.set_num_threads) and [`torch.set_num_interop_threads(int)`](https://pytorch.org/docs/stable/generated/torch.set_num_interop_threads.html#torch.set_num_interop_threads)). Therefore, if the implementation is done for TF, it will likely not be totally lost afterwards as some parts may be useful for PyTorch.

**Comment by kousu** [comment]
I think Taowa was just asking for us to change the line to something like:

```
     jobs = int(math.ceil(multiprocessing.cpu_count()*0.98))
```

**Comment by joshuacwnewton** [comment]
> I think Taowa was just asking for us to change the line to something like:
> 
> ```
>      jobs = int(math.ceil(multiprocessing.cpu_count()*0.98))
> ```

The trouble here is... the value of `jobs` has no bearing on the CPU usage of SCT's scripts. Regardless of how we set `jobs`, TensorFlow will consume all of the resources it has available to it unless we apply a fix like what was suggested by @EvanBeal in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3620#issuecomment-990420725 . :sweat:

**Comment by kousu** [comment]
I just remembered there was some profiling of this script done in our infra repo, motivated by the same observation that we were overflowing our internal infra: https://github.com/neuropoly/computers/issues/84

> As far as I understand, regardless of how we set jobs, TensorFlow will consume all of the resources it has available to it unless we apply a fix like what was suggested

Oh I see. Okay, well forget tensorflow then.

**Comment by jcohenadad** [comment]
SCT is no more relying on TF, so this issue is _possibly_ outdated. If it comes back, please feel free to reopen!

**Comment by kousu** [comment]
Also I've just tempered this as far as our internal clusters are concerned by using [systemd's resource control](https://manpages.ubuntu.com/manpages/jammy/en/man5/systemd.resource-control.5.html): https://github.com/neuropoly/computers/commit/fc11d1a7ff923c994a7100073b37a2e62b8e8a3e. This doesn't help anyone else using SCT on large clusters, but it's probably Good Enough.
